bots/task/distributed_queue.py

```diff
@@ -0,0 +1,75 @@
+# This file is part of Cockpit.
+#
+# Copyright (C) 2019 Red Hat, Inc.
+#
+# Cockpit is free software; you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation; either version 2.1 of the License, or
+# (at your option) any later version.
+#
+# Cockpit is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+
+# Shared GitHub code. When run as a script, we print out info about
+# our GitHub interacition.
+
+import ssl
+
+import pika
+
+__all__ = (
+    'DistributedQueue',
+    'BASELINE_PRIORITY',
+    'MAX_PRIORITY',
+)
+
+BASELINE_PRIORITY = 5
+MAX_PRIORITY = 9
+
+arguments = {
+    'rhel': {
+        "x-max-priority": MAX_PRIORITY
+    },
+    'public': {
+        "x-max-priority": MAX_PRIORITY
+    },
+}
+
+class DistributedQueue(object):
+    # amqp_server should be formatted as host:port
+    # queues should be a list of strings with the names of queues
+    def __init__(self, amqp_server, queues, **kwargs):
+        try:
+            host, port = amqp_server.split(':')
+        except ValueError:
+            raise Exception('Please format amqp_server as host:port')
+        context = ssl.create_default_context(cafile='/run/secrets/webhook/ca.pem')
+        context.load_cert_chain(keyfile='/run/secrets/webhook/amqp-client.key',
+            certfile='/run/secrets/webhook/amqp-client.pem')
+        context.check_hostname = False
+        self.connection = pika.BlockingConnection(pika.ConnectionParameters(
+            host=host,
+            port=int(port),
+            ssl_options=pika.SSLOptions(context, server_hostname=host),
+            credentials=pika.credentials.ExternalCredentials()))
+        self.channel = self.connection.channel()
+        self.declare_results = {}
+
+        for queue in queues:
+            try:
+                result = self.channel.queue_declare(queue=queue, arguments=arguments.get(queue, None), **kwargs)
+                self.declare_results[queue] = result
+            except pika.exceptions.ChannelClosedByBroker as e:
```

Comment:
Is this an "expected"  exception? If not, then it's fine to just propagate it, otherwise the documentation should say what happens to non-existing queues.

Reply:
This is indeed expected, but I added a reply_code check as well. We only recognize this exception with 404. Anything else I'll propagate.